### PR TITLE
Don't try to load all dictionaries with enchant

### PR DESCRIPTION
**Pull request by ilya-fedin**
On debian enchant compiled without hspell, but on others with hspell. Because the current algorithm adds all dictionaries (which is incorrect), on non-debian distros, when enchant tries to check a word with hspell - it getitng an error.
Fixes telegramdesktop/tdesktop#6907, telegramdesktop/tdesktop#6983 (and possible duplicate telegramdesktop/tdesktop#6922)

### CODE REVIEW

**Comment by farseerfc** [comment]
Is this change only loading "en" dictionaries and ignoring all other dicts people may have installed in their system?
I understand that loading all dicts is bad for memory usage and choosing by locale settings is bad for multi-lingual people, and it may require UI rework for users to choose dicts, but I don't think forcing one language is the solution.
Maybe we can blacklist the hebrew/hspell backend temporarily instead, and work on a better way to choose dicts.

**Comment by 23rd** [comment]
> Is this change only loading "en" dictionaries and ignoring all other dicts people may have installed in their system?

Not only the English dictionary, but also the primary system language dictionary (although the check if "English = system language" is missed).

> I understand that loading all dicts is bad for memory usage and choosing by locale settings is bad for multi-lingual people, and it may require UI rework for users to choose dicts, but I don't think forcing one language is the solution.

I agree with you, if I can accept that solution, then only as a temporary one.

**Comment by farseerfc** [comment]
> Not only the English dictionary, but also the primary system language dictionary (although the check if "English = system language" is missed).

Sorry I missed that part. Now this seems acceptable as a temporary one.
Still I want to suggest blacklisting hspell provider because Hebrew users may want English spellchecker.

**Comment by ilya-fedin** [comment]
> Maybe we can blacklist the hebrew/hspell backend temporarily instead, and work on a better way to choose dicts.

There are also issue with memory usage.


>  but I don't think forcing one language is the solution.

I see only one permanent solution - add the ability to choose which dictionaries to use. I can try to implement this, but I'm not sure if this will be accepted.

**Comment by ilya-fedin** [comment]
> Is this change only loading "en" dictionaries and ignoring all other dicts people may have installed in their system?

Only a dictionary that matches the system language, or English if the first isn't found.

**Comment by 23rd** [comment]
> Only a dictionary that matches the system language, or English if the first isn't found.

Oh, yes, I misread diffs from the `File changed` tab.

> I see only one permanent solution - add the ability to choose which dictionaries to use in the GUI. I can try to implement this, but I'm not sure if this will be accepted.

Well, I planned a GUI language picker as part of integration of the custom library for [platforms that don't have the system spell checker](https://github.com/telegramdesktop/tdesktop/issues/96#issuecomment-550201078) before `@mymedia2` added Linux support.

**Comment by ilya-fedin** [comment]
> Well, I planned a GUI language picker as part of integration of the custom library

And you started to write?

File: spellcheck/platform/linux/spellcheck_linux.cpp [hunk 1]
```diff
@@ -33,35 +33,22 @@ class EnchantSpellChecker {
 	EnchantSpellChecker(const EnchantSpellChecker&) = delete;
 	EnchantSpellChecker& operator =(const EnchantSpellChecker&) = delete;
 
+	using DictPtr = std::unique_ptr<enchant::Dict>;
+
 	std::unique_ptr<enchant::Broker> _brokerHandle;
-	std::vector<std::unique_ptr<enchant::Dict>> _validators;
+	std::vector<DictPtr> _validators;
```
Comment (mymedia2) [inline spellcheck/platform/linux/spellcheck_linux.cpp]: Here you actually have the vector that contains no more than one validator.

Comment (ilya-fedin) [reply spellcheck/platform/linux/spellcheck_linux.cpp]: _temporary_
Does it make sense to remove the vector when this is a temporary solution?

**Comment by ilya-fedin** [comment]
I added language detection with Sonnet which can be used as a permanent solution
See also:
https://github.com/desktop-app/cmake_helpers/pull/9
https://github.com/telegramdesktop/tdesktop/pull/6993

File: spellcheck/platform/linux/spellcheck_linux.cpp [hunk 2]
```diff
@@ -146,7 +166,8 @@ void Init() {
 }
 
 bool IsAvailable() {
-	return EnchantSpellChecker::instance()->isAvailable();
+	static auto Available = enchant::loader::do_explicit_linking();
+	return Available;
```
Comment (23rd) [inline spellcheck/platform/linux/spellcheck_linux.cpp]: @ilya-fedin I think I can accept these changes if you put them on a separate commit. (WIth removing of the `EnchantSpellChecker::isAvailable` method.)

**Pull request closed without merging**